src/Compiler/Compile.hs

```diff
@@ -322,34 +332,32 @@ makeRelativeToPaths paths fname
        (root,stem)
 
 
-compileModule :: Terminal -> Flags -> Modules -> Name -> IO (Error Loaded)
-compileModule term flags modules name  -- todo: take force into account
+compileModule :: Terminal -> Flags -> Modules -> Modules -> Name -> CompileTarget () -> IO (Error (Loaded, Maybe FilePath))
+compileModule term flags modules cachedModules name compileTarget-- todo: take force into account
   = runIOErr $
     do let imp = ImpProgram (Import name name rangeNull Private)
-       loaded <- resolveImports name term flags "" initialLoaded{ loadedModules = modules } [imp]
+       loaded <- resolveImports compileTarget (const Nothing) name term flags "" initialLoaded{ loadedModules = modules } cachedModules [imp]
        -- trace ("compileModule: loaded modules: " ++ show (map modName (loadedModules loaded))) $ return ()
        case filter (\m -> modName m == name) (loadedModules loaded) of
-         (mod:_) -> return loaded{ loadedModule = mod }
+         (mod:_) -> return (loaded{ loadedModule = mod }, Nothing)
          []      -> fail $ "Compiler.Compile.compileModule: module not found in imports: " ++ show name ++ " not in " ++ show (map (show . modName) (loadedModules loaded))
 
 
 {---------------------------------------------------------------
   Internal compilation
 ---------------------------------------------------------------}
-compileProgram :: Terminal -> Flags -> Modules -> CompileTarget () -> FilePath -> UserProgram -> IO (Error Loaded)
-compileProgram term flags modules compileTarget fname program
-  = runIOErr $ compileProgram' term flags modules compileTarget  fname program
-
+compileProgram :: Terminal -> Flags -> Modules -> Modules -> CompileTarget () -> FilePath -> UserProgram -> IO (Error (Loaded, Maybe FilePath))
+compileProgram term flags modules cachedModules compileTarget fname program
+  = runIOErr $ compileProgram' (const Nothing) term flags modules cachedModules compileTarget fname program
 
-compileProgramFromFile :: Terminal -> Flags -> Modules -> CompileTarget () -> FilePath -> FilePath -> IOErr Loaded
-compileProgramFromFile term flags modules compileTarget rootPath stem
+compileProgramFromFile :: (FilePath -> Maybe (BString, FileTime)) -> Maybe BString -> Terminal -> Flags -> Modules -> Modules -> CompileTarget () -> FilePath -> FilePath -> IOErr (Loaded, Maybe FilePath)
```

Comment:
The main difference from before in compiling a file:

1. There is a parameter `maybeContents` that get's threaded through, which gives in-memory representation of the file (from the editor). This is used instead of the file on disk (it has priority). The virtual file system only keeps files that are open in the editor alive. 
2. There is a parameter `cachedModules` that has modules that have been compiled previously. We can skip inference / typechecking / persisting these. (We do have to keep track of when they do need to actually be persisted - for example when we request the executable)
3. There is a parameter `contents`, that provides the currently selected program's contents according to `maybeContents`, this parameter could be dropped and just use a local variable, I might do that.
4. Compiling returns the filepath of the compiled module in addition to loaded. This allows for us to know the executable path to report to the editor for running the produced executable.